File: src/main/kotlin/data_design/01_ProductTypes.kt
package data_design

/** Product types
 Given types A and B their product type is A x B
    -> Possible instances are all combination of A and B
 Other name: Record Type
 Similar to classical database tables (one column for each type)
 */
object ProductTypesSimple {
    val pair: Pair<String, Int>             = Pair("Nabil", 33)
    val triple: Triple<String, Int, String> = Triple("Nabil", 33, "Rewe Digital")

    val myName = pair.first // "Nabil"
}

object ProductTypesAdvanced {
    data class Person(val name: String, val age: Int, val workingPlace: String)

    val nabil: Person = Person("Nabil", 33, "Rewe Digital")
    val myName        = nabil.name // "Nabil"
}

fun main() {

    println("--- simple ---")
    println("pair = ${ProductTypesSimple.pair}")
    println("triple = ${ProductTypesSimple.triple}")
    println("myName = ${ProductTypesSimple.myName}")

    println("--- advanced ---")
    println("nabil = ${ProductTypesAdvanced.nabil}")
    println("myName = ${ProductTypesAdvanced.myName}")

}

File: src/main/scala/data_design/01_ProductTypes.scala
package data_design

/** possible values |X| = |A| * |B| */
object ProductTypesSimple:
  val tuple: (String, Int)          = ("Nabil", 36)
  val triple: (String, Int, String) = ("Nabil", 36, "Rewe Digital")

  // Scala tuples of arbitry length
  val tupleCombined: (String, String, Int, String, Int, String) =
    "Combined" *: tuple ++ triple

  val myName = tuple._1 // "Nabil"

object ProductTypesAdvanced:

  case class Person(name: String, age: Int, workingPlace: String)

  val nabil: Person = Person("Nabil", 36, "Rewe Digital")
  val myName        = nabil.name // "Nabil"

@main()
def productTypes =
  println("--- simple ---")
  // String interpolation -> s""
  println(s"pair = ${ProductTypesSimple.tuple}")
  println(s"triple = ${ProductTypesSimple.triple}")
  println(s"triple = ${ProductTypesSimple.tupleCombined}")
  println(s"myName = ${ProductTypesSimple.myName}")

  println("--- advanced ---")
  println(s"nabil = ${ProductTypesAdvanced.nabil}")
  println(s"myName = ${ProductTypesAdvanced.myName}")
